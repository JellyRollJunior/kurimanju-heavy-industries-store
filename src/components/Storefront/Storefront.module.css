.layout {
  composes: primaryBackground from '../../styles/colors.module.css';
  composes: secondaryColor from '../../styles/colors.module.css';
  display: flex;
  flex-direction: column;
  min-height: 100vh;
}

main {
  padding: 12px;
}

.shopHeader {
  display: flex;
  justify-content: center;

  --header-color: #ffcc77;
  background: repeating-linear-gradient(
    45deg,
    transparent,
    transparent 10px,
    var(--header-color) 10px,
    var(--header-color) 20px
  );
}

.shopHeaderTitle {
  composes: primaryBackground from '../../styles/colors.module.css';
  composes: quaternaryColor from '../../styles/colors.module.css';
  padding: 3px 12px 0 12px;
  font-size: 20px;
  border: 2px solid;
}

.shop {
  composes: quaternaryColor from '../../styles/colors.module.css';
  display: flex;
  flex-direction: column;
  gap: 28px;
  padding: 42px 24px 24px 24px;

  border: 2px solid;
  border-top: none;
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
}

@media (min-width: 760px) {
  main {
    padding: 0 12px 12px 12px;
  }

  .layout {
    display: grid;
    grid-template-columns: 320px 1fr;
  }
}

@media (min-width: 1000px) {
  .shop {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
  }
}

@media (min-width: 1400px) {
  .shop {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
  }
}

.cartOverlay {
  display: flex;
  justify-content: end;
  padding: 36px 0;

  position: fixed;
  right: 0;
  height: 100vh;
  width: 100vw;
  background-color: #00000080;
}

.cart {
  composes: primaryBackground from '../../styles/colors.module.css';
  composes: quaternaryColor from '../../styles/colors.module.css';
  min-width: 250px;

  text-align: center;
  border-top: 2px solid;
  border-left: 2px solid;
  border-bottom: 2px solid;
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
}