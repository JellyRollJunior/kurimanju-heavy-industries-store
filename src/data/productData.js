import x4_base from '../assets/x4_base.png';
import x4_stealth from '../assets/x4_stealth.png';
import x4_respond from '../assets/x4_respond.png';
import x4_enforce from '../assets/x4_enforce.png';
import x4_supress from '../assets/x4_supress.png';

const productData = [
    {
        id: '0',
        name: 'X-4 Base',
        price: '90 000 000',
        img: x4_base,
    },
    {
        id: '1',
        name: 'X-4 Stealth',
        price: '95 000 000',
        img: x4_stealth,
    },
    {
        id: '2',
        name: 'X-4 Respond',
        price: '95 000 000',
        img: x4_respond,
    },
    {
        id: '3',
        name: 'X-4 Enforce',
        price: '100 000 000',
        img: x4_enforce,
    },
    {
        id: '4',
        name: 'X-4 Supress',
        price: '120 000 000',
        img: x4_supress,
    },
];

export { productData };
