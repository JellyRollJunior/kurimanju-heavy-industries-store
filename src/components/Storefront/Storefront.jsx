import { useState } from 'react';
import { productData } from '../../data/productData.js';
import { Header } from '../Header/Header.jsx';
import { Card } from '../Card/Card.jsx';
import { Cart } from '../Cart/Cart.jsx';
import styles from './Storefront.module.css';
import cartAnimations from '../Cart/Cart.module.css';

const Storefront = () => {
  // const [cart, setCart] = useState(Array(productData.length).fill(0));
  const [cart, setCart] = useState([1, 1, 1, 1, 1]);
  const MAX_QUANTITY_PER_ITEM = 10;

  const addToCart = (id, quantity) => {
    const newCart = [...cart];
    newCart[id] = Math.min(MAX_QUANTITY_PER_ITEM, newCart[id] + Number(quantity));
    setCart(newCart);
  };

  const removeFromCart = (id) => {
    const newCart = [...cart];
    newCart[id] = 0;
    setCart(newCart);
  };

  const checkoutCart = () => {
    alert('Checkout successful');
    hideCart();
    setCart(Array(productData.length).fill(0));
  };

  const displayCart = () => {
    const cart = document.querySelector('aside');
    document.body.style.overflow = 'hidden';
    cart.style.display = 'block';
    setTimeout(() => {
      cart.classList.add(`${cartAnimations.slideIn}`);
      cart.classList.remove(`${cartAnimations.slideOut}`);
    }, 0);
  };

  const hideCart = () => {
    const cart = document.querySelector('aside');
    document.body.style.overflow = '';
    cart.classList.add(`${cartAnimations.slideOut}`);
    cart.classList.remove(`${cartAnimations.slideIn}`);
    setTimeout(() => {
      cart.style.display = 'none';
    }, 1000);
  };

  return (
    <div className={styles.layout}>
      <Header onClickViewCart={displayCart} />
      <main>
        <header className={styles.shopHeader}>
          <h2 className={styles.shopHeaderTitle}>SHOP</h2>
        </header>
        <section className={styles.shop}>
          {productData.map((item) => (
            <Card
              key={item.id}
              id={item.id}
              name={item.name}
              price={item.price}
              img={item.img}
              maxQuantityPerItem={MAX_QUANTITY_PER_ITEM}
              onSubmit={addToCart}
            ></Card>
          ))}
        </section>
      </main>
      <Cart
        cart={cart}
        hideCart={hideCart}
        checkoutCart={checkoutCart}
        removeFromCart={removeFromCart}
      />
    </div>
  );
};

export { Storefront };
