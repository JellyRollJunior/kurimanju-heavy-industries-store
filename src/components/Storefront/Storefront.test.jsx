import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Storefront } from './Storefront';
import { productData } from '../../data/productData.js';

vi.mock('../Header/Header', () => ({
  Header: () => <header>I am a header</header>
}));

vi.mock('../Card/Card', () => ({
  Card: () => <div data-testid='card'>I am a card</div>
}))

vi.mock('../Footer/Footer', () => ({
  Footer: () => <footer>I am a footer</footer>
}))

vi.mock('../Cart/Cart', () => ({
  Cart: () => <aside>I am a cart</aside>
}))

describe('Storefront component', () => {
  it('Renders header, shop, shop items, and cart', () => {
    const { container } = render(<Storefront />);

    expect(container).toMatchSnapshot();
  });

  it('Renders a card for each product', () => {
    render(<Storefront />);

    expect(screen.getAllByTestId('card')).toHaveLength(productData.length);
  })
});
