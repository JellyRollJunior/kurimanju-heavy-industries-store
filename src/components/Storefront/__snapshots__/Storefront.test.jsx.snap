// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Storefront component > Renders header, shop, shop items, and cart 1`] = `
<div>
  <div
    class="_layout_5546c3"
  >
    <header>
      I am a header
    </header>
    <main>
      <header
        class="_shopHeader_5546c3"
      >
        <h2
          class="_shopHeaderTitle_5546c3"
        >
          SHOP
        </h2>
      </header>
      <section
        class="_shop_5546c3"
      >
        <div
          data-testid="card"
        >
          I am a card
        </div>
        <div
          data-testid="card"
        >
          I am a card
        </div>
        <div
          data-testid="card"
        >
          I am a card
        </div>
        <div
          data-testid="card"
        >
          I am a card
        </div>
        <div
          data-testid="card"
        >
          I am a card
        </div>
      </section>
    </main>
    <footer>
      I am a footer
    </footer>
    <aside>
      I am a cart
    </aside>
  </div>
</div>
`;
